backend/src/v2/driver/driver.go

```diff
@@ -1188,43 +1198,311 @@ func resolveInputs(ctx context.Context, dag *metadata.DAG, iterationIndex *int,
-			taskOutput := artifactSpec.GetTaskOutputArtifact()
-			if taskOutput.GetProducerTask() == "" {
-				return nil, artifactError(fmt.Errorf("producer task is empty"))
+			cfg := resolveUpstreamArtifactsConfig{
+				ctx:           ctx,
+				artifactSpec:  artifactSpec,
+				dag:           dag,
+				pipeline:      pipeline,
+				mlmd:          mlmd,
+				inputs:        inputs,
+				name:          name,
+				artifactError: artifactError,
 			}
-			if taskOutput.GetOutputArtifactKey() == "" {
-				return nil, artifactError(fmt.Errorf("output artifact key is empty"))
+			if err := resolveUpstreamArtifacts(cfg); err != nil {
+				return nil, err
 			}
-			tasks, err := getDAGTasks()
+		default:
+			return nil, artifactError(fmt.Errorf("artifact spec of type %T not implemented yet", t))
+		}
+	}
+	// TODO(Bobgy): validate executor inputs match component inputs definition
+	return inputs, nil
+}
+
+// getDAGTasks is a recursive function that returns a map of all tasks across all DAGs in the context of nested DAGs.
+func getDAGTasks(
+	ctx context.Context,
+	dag *metadata.DAG,
+	pipeline *metadata.Pipeline,
+	mlmd *metadata.Client,
+	flattenedTasks map[string]*metadata.Execution,
+) (map[string]*metadata.Execution, error) {
+	if flattenedTasks == nil {
+		flattenedTasks = make(map[string]*metadata.Execution)
+	}
+	currentExecutionTasks, err := mlmd.GetExecutionsInDAG(ctx, dag, pipeline, true)
+	if err != nil {
+		return nil, err
+	}
+	for k, v := range currentExecutionTasks {
+		flattenedTasks[k] = v
+	}
+	for _, v := range currentExecutionTasks {
+
+		if v.GetExecution().GetType() == "system.DAGExecution" {
+			// Iteration count is only applied when using ParallelFor, and in
+			// that scenario you're guaranteed to have redundant task names even
+			// within a single DAG, which results in an error when
+			// mlmd.GetExecutionsInDAG is called. ParallelFor outputs should be
+			// handled with dsl.Collected.
+			_, ok := v.GetExecution().GetCustomProperties()["iteration_count"]
+			if ok {
+				glog.Infof("Found a ParallelFor task, %v. Skipping it.", v.TaskName())
+				continue
+			}
+			glog.V(4).Infof("Found a task, %v, with an execution type of system.DAGExecution. Adding its tasks to the task list.", v.TaskName())
+			subDAG, err := mlmd.GetDAG(ctx, v.GetExecution().GetId())
+			if err != nil {
+				return nil, err
+			}
+			// Pass the subDAG into a recursive call to getDAGTasks and update
+			// tasks to include the subDAG's tasks.
+			flattenedTasks, err = getDAGTasks(ctx, subDAG, pipeline, mlmd, flattenedTasks)
 			if err != nil {
-				return nil, artifactError(err)
+				return nil, err
 			}
-			producer, ok := tasks[taskOutput.GetProducerTask()]
+		}
+	}
+
+	return flattenedTasks, nil
+}
+
+// resolveUpstreamParametersConfig is just a config struct used to store the
+// input parameters of the resolveUpstreamParameters function.
+type resolveUpstreamParametersConfig struct {
```

Comment:
Do you see opportunities to consolidate and merge this with `resolveUpstreamArtifacts...`?
I tend to think the logic should be very similar between the two?

Reply:
Great question! We wondered the same thing (and even tried it out) but the resolution logic is so different that it results in almost no deduplication at the cost of a massive, convoluted function full of nested conditionals. 

That being said, your comment inspired us to consolidate the two config structs that are passed to the two functions. Thanks!